### DISCUSSION TITLE
Injeção de dependência Blazor Net 8

### DISCUSSION THREAD

**Discussion by Shecla** [balta-io/forum#464] [Tira Dúvidas] [answered]
Tenho uma classe com várias propriedades. Defino os dados das propriedades no componente de login (Account/Login) no lado SERVER utilizando a injeção por dependência. Como faço para ler os valores das propriedades desta mesma classe no lado CLIENT?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @Shecla, como vai?

Poderia por favor colocar os códigos que produziu? Assim fica mais fácil entender o que precisa...
Detalha os trechos de código, o que precisa que aconteça...

**Comment by Shecla** [comment]
Criei um projeto que desejo compartilhar uma classe (dadosemcomum.cs) por entre o server e o client. Faço a injeção no program.cs tanto do lado server como no client. Defino a propriedade da classe no Home.razor, que esta no server. Quando chamo esta mesma propriedade no Count.rzor, que esta no lado client, ela vem nula.
[BlazorApp8.zip](https://github.com/balta-io/forum/files/13942108/BlazorApp8.zip)

**Reply by andrebaltieri** [reply]
Depende... é uma classe sem nenhuma dependência Web (ASP.NET, Razor Blazor)? É uma classe para passagem de dados ou para execução de algum serviço?

Dependendo do que quiser fazer e como seu código estiver criado, será um tipo de referência, inclusive talvez nem precise de DI!

**Reply by Shecla** [reply]
É uma classe para passagem de dados. Na tela de login acesso uma API externa que checa a autenticação do usuário e gera um token que será utilizado nas chamadas das API. Como só tenho a senha do uauário na tela de login, não em como gerar este token apenas do lado CLIENT.

**Reply by Shecla** [reply]
Enviei este exemplo, ilustrativo, para mostrar que preciso definir os dados de uma classe do lado Server e enviar estes para o client.

**Comment by Shecla** [comment]
Existe alguma forma de passar os dados entre os projetos Server e Client? Como fazer?

**Reply by malighieri** [reply]
Olá. Vi sua solução com seus dois projetos, sendo um nomeado como BlazorApp8 e o outro como BlazorApp8.Client.
Acredito que você se equivocou ao criar estes dois projetos desta forma. 
Em um projeto web usando Blazor, você tem as duas camadas no mesmo projeto. Ou seja, você não precisa de um segundo projeto cliente pois o seu projeto Blazor pode ter o backend e o frontend juntos. E assim não teria essa questão de compartilhar informações entre os projetos, pois só existiria um projeto.

Nos diga se sua intenção era realmente ter dois projetos, um cliente e outro servidor para que possamos tentar ajudar mais, e direcionar você para o melhor caminho.

@andrebaltieri , caso eu tenha falado alguma besteira, por favor me corrija.
Abraços!

**Reply by Shecla** [reply]
Acho que as novas aplicações do Blazor no Net 8 sempre são criadas com os dois projetos server e client.

**Reply by andrebaltieri** [reply]
Pessoal, sobre essa segmentação, eu explico tudo no curso de Blazor 8 com .NET 8 e C# 12:
https://balta.io/cursos/fundamentos-blazor-dotnet-8

Minha recomendação é seguir todos os cursos da carreira de Blazor:
https://balta.io/carreiras/aspnet-blazor-developer

**Reply by Shecla** [reply]
Vi os cursos, mas mesmo assim ainda não sei como fazer para passar as informações que são definidas no projeto Server para o projeto Client, como esta no aplicativo que enviei anteriormente.

**Reply by andrebaltieri** [reply]
@Shecla uma saída é criar um terceiro projeto que compartilha estes objetos, mas o projeto server referencia o cliente até onde sei.
Coloca os trechos de código, explica eles, detalha certinho o que você precisa que aconteça... assim conseguimos ajudar melhor.

**Comment by Shecla** [comment]
Criei este projeto
![image](https://github.com/balta-io/forum/assets/16648349/412ad3b4-01f5-41c3-8e2e-0eafde12c928)

**Faço esta DI no program.cs do server e do client**
builder.Services.AddScoped<DadosEmComum>();

**No Home.razor, que esta no projeto server defino a classe**
 protected override void OnInitialized()
 {
     dados.NomeUsuario = "CLAUDIO";
 }

**no Counter.razor**
@page "/counter"
@* @rendermode InteractiveAuto *@
 

@inject DadosEmComum dados

@rendermode @(new InteractiveAutoRenderMode(prerender: false))

<PageTitle>Counter</PageTitle>

<h1>Counter</h1>

<h1>@dados.NomeUsuario</h1>
<h1>@nomeUsuario</h1>

<p role="status">Current count: @currentCount</p>

<button class="btn btn-primary" @onclick="IncrementCount">Click me</button>

@code {
    private int currentCount = 0;
    private string? nomeUsuario = "";

    private void IncrementCount()
    {
        currentCount++;
    }

    protected override void OnInitialized()
    {
        nomeUsuario = dados.NomeUsuario;
    }
}

O que eu quero é que os dados da classe, "DadosEmComum.cs" que foram definidos no server, já estejam disponíveis no Client ao chamar a página Counter

**Comment by Shecla** [comment]
Esta é a DI que faço 
builder.Services.AddScoped<DadosEmComum>();

**Comment by Shecla** [comment]
![image](https://github.com/balta-io/forum/assets/16648349/a4ba9b82-b129-43e5-8340-29540d8c6948)

**Comment by Shecla** [comment]
[BlazorApp8.zip](https://github.com/balta-io/forum/files/13989462/BlazorApp8.zip)

**Comment by andrebaltieri** [comment]
Bom dia, @Shecla, como vai?
Pelo que vi é isto mesmo, os dados estão sendo preenchidos mas não exibem?

Recomendo também revisar nosso conteúdo sobre injeção de dependência:
https://balta.io/cursos/dominando-injecao-de-dependencia

**Reply by Shecla** [reply]
**Pelo que vi é isto mesmo, os dados estão sendo preenchidos mas não exibem?**
Exatamente este é o meu problema.

**Reply by Shecla** [reply]
Então não tem solução para este procedimento onde defino os dados no "server project" e ler no "client project" usando DI?

**Reply by andrebaltieri** [reply]
Bom dia, @Shecla, como vai? Os dados estão sendo preenchidos antes do componente que está no cliente chamar ele? Você consegue depurar pra ver?

Não tem como passar essa informação já no componente como parâmetro pra não precisar fazer esse processo de injetar os itens no client?

Tudo que está no Client fica exposto no Browser, como comento nos cursos de Blazor e Blazor WASM, então tem que tomar muito cuidado com o que você referencia ou injeta ali.

**Reply by Shecla** [reply]
**Bom dia, @Shecla, como vai? Os dados estão sendo preenchidos antes do componente que está no cliente chamar ele? Você consegue depurar pra ver?**
Sim, como no app que enviei, os dados são preenchidos na Home e visualizados no Counter

**Não tem como passar essa informação já no componente como parâmetro pra não precisar fazer esse processo de injetar os itens no client?**
Nesta aplicação, sim, mas na aplicação real teria que passar uma classe usando o "NavigationManager.NavigateTo" como parâmetro e os dados ficariam visíveis na URL, não?

**Tudo que está no Client fica exposto no Browser, como comento nos cursos de Blazor e Blazor WASM, então tem que tomar muito cuidado com o que você referencia ou injeta ali.**
Neste exemplo não teria problema, mas como guardar o token que utilizo para chamar as API externas, no Blazor WASM?

**Reply by andrebaltieri** [reply]
Bom dia, @Shecla então aqui está seu problema:
"Sim, como no app que enviei, os dados são preenchidos na Home e visualizados no Counter"

No caso, quando navegamos o estado se perde, ele não é mantido e se você usar o Singleton ele será mantido na memória, de forma compartilhada.

Existem várias formas de gerenciar o estado, uma delas é esta aqui:
https://learn.microsoft.com/en-us/aspnet/core/blazor/state-management?view=aspnetcore-8.0&pivots=server

Sobre o Token, você pode armazenar no LocalStorage e ler quando precisar, ou usar Cookies que são enviados em todas as requisições.
